### PR TITLE
VY-RES-456 - API for available resources

### PR DESCRIPTION
**Pull request by lf-achyutpkl**
@lf-kailashraj @anishmanandhar @krishnatimilsina please review my code

### CODE REVIEW

File: src/main/java/com/lftechnology/vyaguta/resource/pojo/Employee.java [hunk 1]
```diff
@@ -55,6 +55,31 @@
     @Transient
     private String gender;
 
+    @Transient
+    private Designation designation;
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/pojo/Employee.java]: CRITICAL Make "designation" transient or serializable. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1948)

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectServiceImpl.java [hunk 2]
```diff
@@ -260,7 +261,7 @@ private void fixContract(Project project) {
         Double totalEmployee = Double.valueOf(employeeService.fetchActiveEmployees().size());
         Double freeResourceCount = totalEmployee - bookedResourcesCount;
 
-        Map<String, Object> resultOutput = new HashMap<String, Object>();
+        Map<String, Object> resultOutput = new HashMap<>();
         resultOutput.put("totalResources", totalEmployee);
         resultOutput.put("bookedResources", new HashMap<String, Double>() {
             {
```
Comment (bhuwang) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectServiceImpl.java]: MAJOR Move the contents of this initializer to a standard constructor or to field initializers. [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1171)

**Comment by bhuwang** [comment]
SonarQube analysis reported 3 issues:
* CRITICAL 1 critical
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal "Invalid date format" 3 times.](https://github.com/leapfrogtechnology/vyaguta-resource/blob/1e423e70a2b6c579e878d29f176962f78f985e96/src/main/java/com/lftechnology/vyaguta/resource/rs/ProjectRs.java#L128) [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1192)

File: src/main/java/com/lftechnology/vyaguta/resource/dao/impl/ContractMemberDaoImpl.java [hunk 3]
```diff
@@ -66,4 +66,19 @@ public Long findBookedResourceCount(LocalDate date) {
...
+    @Override
+    public Map<UUID, Double> findAvailableResource(LocalDate date) {
+        Map<UUID, Double> map = new HashMap<>();
+
+        List<Object[]> result = em
+                .createQuery(
+                        "SELECT employee.id, SUM(allocation * 0.01) FROM ContractMember WHERE :date BETWEEN joinDate AND endDate GROUP BY employee.id")
+                .setParameter("date", date).getResultList();
+        for (Object[] obj : result) {
+            map.put((UUID) obj[0], (Double) obj[1]);
```
Comment (lf-kailashraj) [inline src/main/java/com/lftechnology/vyaguta/resource/dao/impl/ContractMemberDaoImpl.java]: ?

File: src/main/java/com/lftechnology/vyaguta/resource/pojo/AvailableResource.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * POJO class holding information about employee and their allocation
+ * 
+ * @author Achyut Pokhrel <achyutpokhrel@lftechnology.com>
+ *
+ */
+public class AvailableResource {
+
+    @Type(type = "pg-uuid")
```
Comment (anishmanandhar) [inline src/main/java/com/lftechnology/vyaguta/resource/pojo/AvailableResource.java]: I don't prefer uuid mapping for specific database. Can you look at this 
https://zorq.net/b/2012/04/21/switching-hibernates-uuid-type-mapping-per-database/

File: src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectServiceImpl.java [hunk 5]
```diff
@@ -271,6 +272,29 @@ private void fixContract(Project project) {
         });
         resultOutput.put("freeResources", freeResourceCount);
         return resultOutput;
+    }
 
+    @Override
+    public List<AvailableResource> findAvailableResource(LocalDate date) {
```
Comment (anishmanandhar) [inline src/main/java/com/lftechnology/vyaguta/resource/service/impl/ProjectServiceImpl.java]: should be findAvailableResources (plural form of naming convention)

**Comment by bhuwang** [comment]
SonarQube analysis reported 2 issues:
* CRITICAL 1 critical
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal "Invalid date format" 3 times.](https://github.com/leapfrogtechnology/vyaguta-resource/blob/3bc53359c484dea625066faaa2d88b30f16b4e7c/src/main/java/com/lftechnology/vyaguta/resource/rs/ProjectRs.java#L128) [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1192)

File: src/main/java/com/lftechnology/vyaguta/resource/pojo/Employee.java [hunk 6]
```diff
@@ -55,6 +55,31 @@
...
+
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class Designation {
```
Comment (lf-kailashraj) [inline src/main/java/com/lftechnology/vyaguta/resource/pojo/Employee.java]: serializable

Comment (anishmanandhar) [inline src/main/java/com/lftechnology/vyaguta/resource/pojo/Employee.java]: why serialisation for static ??

**Comment by bhuwang** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal "Invalid date format" 3 times.](https://github.com/leapfrogtechnology/vyaguta-resource/blob/60f3506a2714af995f28cde09728740f3a753707/src/main/java/com/lftechnology/vyaguta/resource/rs/ProjectRs.java#L128) [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1192)

**Comment by bhuwang** [comment]
SonarQube analysis reported 1 issue:
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MINOR [Define a constant instead of duplicating this literal "Invalid date format" 3 times.](https://github.com/leapfrogtechnology/vyaguta-resource/blob/b5093f07b32f04df4beba4282965a17665fa7d87/src/main/java/com/lftechnology/vyaguta/resource/rs/ProjectRs.java#L128) [rule](http://sonar.java.lftechnology.com:9000/coding_rules#rule_key=squid%3AS1192)

**Pull request merged**